app/src/main/java/com/nutomic/syncthingandroid/views/DevicesAdapter.java

```diff
@@ -45,25 +45,39 @@ public View getView(int position, View convertView, @NonNull ViewGroup parent) {
 
         name.setText(getItem(position).getDisplayName());
         Resources r = getContext().getResources();
-        boolean haveInfo = mConnections != null && mConnections.connections.containsKey(deviceId) &&
-                mConnections.connections.get(deviceId).connected;
-        if (haveInfo) {
-            Connections.Connection conn = mConnections.connections.get(deviceId);
-            if (conn.completion == 100) {
-                status.setText(r.getString(R.string.device_up_to_date));
-                status.setTextColor(ContextCompat.getColor(getContext(), R.color.text_green));
-            }
-            else {
-                status.setText(r.getString(R.string.device_syncing, conn.completion));
-                status.setTextColor(ContextCompat.getColor(getContext(), R.color.text_blue));
+
+        Connections.Connection conn = null;
+        if (mConnections != null) {
+            if (mConnections.connections.containsKey(deviceId)) {
+                conn = mConnections.connections.get(deviceId);
             }
-            download.setText(Util.readableTransferRate(getContext(), conn.inBits));
-            upload.setText(Util.readableTransferRate(getContext(), conn.outBits));
         }
-        else {
+
+        if (conn != null)
+        {
+            download.setText(Util.readableTransferRate(getContext(), (conn.paused || !conn.connected) ? 0 : conn.inBits));
+            upload.setText(Util.readableTransferRate(getContext(), (conn.paused || !conn.connected) ? 0 : conn.outBits));
+
+            if (conn.paused) {
+                status.setText(r.getString(R.string.device_paused));
+                status.setTextColor(ContextCompat.getColor(getContext(), R.color.text_black));
+            } else if (conn.connected) {
+                if (conn.completion == 100) {
+                    status.setText(r.getString(R.string.device_up_to_date));
+                    status.setTextColor(ContextCompat.getColor(getContext(), R.color.text_green));
+                }
+                else {
+                    status.setText(r.getString(R.string.device_syncing, conn.completion));
+                    status.setTextColor(ContextCompat.getColor(getContext(), R.color.text_blue));
+                }
+            } else {
+                status.setText(r.getString(R.string.device_disconnected));
+                status.setTextColor(ContextCompat.getColor(getContext(), R.color.text_red));
+            }
+        } else {
             download.setText(Util.readableTransferRate(getContext(), 0));
             upload.setText(Util.readableTransferRate(getContext(), 0));
-            status.setText(r.getString(R.string.device_disconnected));
+            status.setText(r.getString(R.string.device_state_unknown));
             status.setTextColor(ContextCompat.getColor(getContext(), R.color.text_red));
         }
```

Comment:
This whole block could be rewritten as:
```
if (conn == null) {
   do stuff
   return early
}

if (conn.paused) {
  do stuff
  return early
}

if (conn.connected) {
  do stuff
  return early
}

set disconnected
return
```

This way there is less nesting and easier to follow.

Reply:
Agreed.